### PR TITLE
Fixed optimisers not applying gradient to parameters

### PR DESCRIPTION
**Pull request by GenaBitu**
Hi,
the more advanced optimisers (all except SGD) weren't applying the computed gradient to the parameters. Such behaviour seems rather confusing to me, so I changed it so that all optimisers behave in the same way.

Thank you,
GenaBitu

### CODE REVIEW

**Comment by MikeInnes** [comment]
Thanks a lot for the patch. The original idea here was that the functions in `optimisers.jl` would serve as building blocks, which you can combine together to form a real optimiser in `interface.jl`. For example you might define ADAM like this (following [Keras' optimisers](https://keras.io/optimizers/)):

```julia
ADAM(ps, η = 0.001, β1 = 0.9, β2 = 0.999, ϵ = 1e-08, decay = 0.0) =
  optimiser(ps, p -> adam(p; ...), p -> invdecay(p, decay), p -> descent(p, 1))
```

So as a user you'd construct `ADAM(ps)` and the `descent` call would actually apply the gradient.

Do you think you could try working things into this form? Let me know if that's not clear at all.

**Comment by pevnak** [comment]
The construction as Mike has suggested works nicely. I would stick with it and add the needed API to it.

**Comment by MikeInnes** [comment]
Thanks a lot for giving this a go -- but for now I'll close this in favour of #70.

**Comment by GenaBitu** [comment]
Agreed, that's a better solution given the original intent. I am just wondering whether `descent` is the right name. That's really minor though...
